Mathlib/CategoryTheory/Products/Bifunctor.lean

```diff
@@ -57,3 +58,72 @@ theorem diagonal' (F : C × D ⥤ E) (X X' : C) (f : X ⟶ X') (Y Y' : D) (g : Y
 #align category_theory.bifunctor.diagonal' CategoryTheory.Bifunctor.diagonal'
 
 end CategoryTheory.Bifunctor
+
+namespace CategoryTheory.Functor
+
+/-- Opposite of a bifunctor.
+-/
+@[simps!]
+def biop (F : Cᵒᵖ ⥤ D ⥤ E) : C ⥤ Dᵒᵖ ⥤ Eᵒᵖ := F.rightOp ⋙ Functor.opHom _ _
+
+end CategoryTheory.Functor
+
+/- Utilities to apply functors into functor categories. -/
+namespace CategoryTheory.Functor
+
+-- We would like to keep these opaque in order to avoid leaking implementation details.
+
+/- Action of two-variable functors on objects. -/
+def obj2 (H : C ⥤ D ⥤ E) (A : C) (B : D) : E := (H.obj A).obj B
+
+/- Action of three-variable functors on objects. -/
+def obj3 (H : C ⥤ D ⥤ E ⥤ F) (A : C) (B : D) (C : E) : F := ((H.obj A).obj B).obj C
+
+variable {D₁ D₂ C₁ C₂ E₁ E₂}
+
+/- Action of two-variable functors on morphisms. -/
+def map2 (H : C ⥤ D ⥤ E) (f : C₁ ⟶ C₂) (g : D₁ ⟶ D₂) :
+    (H.obj2 C₁ D₁ ⟶ H.obj2 C₂ D₂) :=
+  (H.map f).app _ ≫ (H.obj C₂).map g
+
+/- Action of three-variable functors on morphisms. -/
+def map3 (H : C ⥤ D ⥤ E ⥤ F) (f : C₁ ⟶ C₂) (g : D₁ ⟶ D₂) (h : E₁ ⟶ E₂) :
+    (H.obj3 C₁ D₁ E₁ ⟶ H.obj3 C₂ D₂ E₂) :=
+  (H.map2 f g).app _ ≫ (H.obj2 C₂ D₂).map h
+
+end CategoryTheory.Functor
+
+/- Natural transformations between functors with many variables. -/
+namespace CategoryTheory.NatTrans
+
+/- Apply a natural transformation between bifunctors to two objects. -/
+def app2 {F G : C ⥤ D ⥤ E} (α : NatTrans F G) (X : C) (Y : D) :
+    F.obj2 X Y ⟶ G.obj2 X Y :=
+  (α.app X).app Y
+
+/- Apply a natural transformation between bifunctors in three variables to three objects. -/
+def app3 {H G : C ⥤ D ⥤ E ⥤ F} (α : NatTrans H G) (X : C) (Y : D) (Z : E) :
+    H.obj3 X Y Z ⟶ G.obj3 X Y Z :=
+  ((α.app X).app Y).app Z
+
+@[simp]
+def naturality2 {H G : C ⥤ D ⥤ E} (α : NatTrans H G) {X Y X' Y'} (f : X ⟶ X') (g : Y ⟶ Y') :
```

Comment:
```suggestion
@[reassoc (attr := simp)]
lemma naturality₂ {H G : C ⥤ D ⥤ E} (α : H ⟶ G) {X Y X' Y'} (f : X ⟶ X') (g : Y ⟶ Y') :
```
Even though they are definitionally equal, I think it is better to use `H ⟶ G` instead of `NatTrans H G`.

Reply:
I agree, done! I can change `naturality₂` and `naturality₃` but I can't do the same for `app₂` and `app₃`, how come?

Reply:
I just noticed that using `H ⟶ G` instead of `NatTrans` does not allow to use `α.naturality₂` as field notation... I guess it's because it needs instance resolution to solve the hom notation, so it does not include the definition as part of the fields of `NatTrans`, or something along these lines.

Reply:
It is pretty awkward to not be able to use dot notation. I've kept the `NatTrans` argument while in the `NatTrans` namespace

Reply:
Then, please check that the simp lemmas work with `H ⟶ G`, and revert to `H ⟶ G` in case it does not.

Reply:
If `map₂` is an abbrev we can't have `naturality₂`, `naturality₃` as simp lemmas (because they aren't in simp normal form). So this is vacuously true but I'm not sure the current state of things (no simp lemmas) is good. I think your suggestion to mark `naturality_app` and `naturality_app_app` as `simp` is a good one, and if we do so the following works (so presumably simp is able to determine that lemmas with a `NatTrans` argument apply to something of type `H ⟶ G`).
```
@[reassoc (attr:=simp)]
theorem naturality_app_app {G H : C ⥤ D ⥤ E ⥤ F} (T : NatTrans G H) (Z : E) 
    {X₁ Y₁ : C} {X₂ Y₂ : D} (f : X₁ ⟶ Y₁) (g : X₂ ⟶ Y₂) :
    (G.map f).app₂ X₂ Z ≫ ((G.obj Y₁).map g).app Z ≫ T.app₃ Y₁ Y₂ Z =
      T.app₃ X₁ X₂ Z ≫ (H.map₂ f g).app Z := by
  rw [naturality_app, ← Category.assoc, ← comp_app₂,
    naturality, comp_app₂, Category.assoc]
  rfl

example {G H : C ⥤ D ⥤ E ⥤ F} (T : G ⟶ H) (Z : E) {X₁ Y₁ : C}
    {X₂ Y₂ : D} (f : X₁ ⟶ Y₁) (g : X₂ ⟶ Y₂) :
    (G.map f).app₂ X₂ Z ≫ ((G.obj Y₁).map g).app Z ≫ T.app₃ Y₁ Y₂ Z =
      T.app₃ X₁ X₂ Z ≫ (H.map₂ f g).app Z := by
  simp
```

Maybe we can plan to mark `naturality_app` as a global simp lemmas after this PR?

Reply:
> Maybe we can plan to mark `naturality_app` as a global simp lemmas after this PR?

I think this could be attempted in this PR.

For the rest, the abbreviations `app₂` and `app₃` are ok, but I think we should essentially get rid of `map₂` and `map₃`, and rely mostly on the already existing automation, with `naturality_app` as a global simp lemma and a new `naturality_app_app` in the following version:
```lean
@[reassoc (attr := simp)]
lemma naturality_app_app
    {G H : C ⥤ D ⥤ E ⥤ F} (T : NatTrans G H) (Y : D) (Z : E) {X X' : C} (f : X ⟶ X') :
    ((G.map f).app Y).app Z ≫ ((T.app X').app Y).app Z =
      ((T.app X).app Y).app Z ≫ ((H.map f).app Y).app Z :=
  congr_app (NatTrans.naturality_app T Y f) Z
```

Then, your version of `naturality_app_app` can be proven by `simp`.

Reply:
I'm building with `naturality_app` marked as simp right now and it doesn't seem to break anything, and I like this version of `naturality_app_app`. However I feel mixed about removing `map₂` and `map₃`. As is they play very poorly with simp, but it's going to be really annoying to write out the definition of `map₂` every time you work with a bifunctor. I think for right now I'll remove them, but if/when we get around to adding profunctors and (co)ends I think this might need to be revisited

Reply:
I've pushed this version of the code. By the way, could `naturality_app`, `naturality_app_app`, `naturality_app_app_app`, ... be a `simproc`? I don't know enough about `simproc`s to say if this would be a good idea